contrib/src/main/java/org/apache/apex/malhar/contrib/kudu/AbstractKuduOutputOperator.java

```diff
@@ -0,0 +1,666 @@
+ *      <li>DELETE</li>
+ * </ol>
+ * </p>
+ * <p>
+ * Please note that the Update mutation allows to change a subset of columns only even though there might be columns
+ * that were defined to be non-nullable. This is because the original mutation of type insert would have written the
+ * mandatory columns. In such scenarios, the method setDoNotWriteColumns() in {@link KuduExecutionContext} can be
+ * used to specify only those columns that need an update. This ways a read and Update pattern can be merged to a
+ * simple update pattern thus avoiding a read if required.</p>
+ * */
+@InterfaceStability.Evolving
+public abstract class AbstractKuduOutputOperator extends BaseOperator
+    implements Operator.ActivationListener<Context.OperatorContext>,Operator.CheckpointNotificationListener
+{
+
+  private ApexKuduConnection apexKuduConnection;
+
+  private transient  KuduTable kuduTable;
+
+  private transient  KuduSession kuduSession;
+
+  private transient KuduClient kuduClientHandle;
+
+  private transient Map<String,ColumnSchema> allColumnDefs;
+
+  private transient Map<String,Object> kuduColumnBasedGetters;
+
+  private Set<String> primaryKeyColumnNames;
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractKuduOutputOperator.class);
+
+  private WindowDataManager windowDataManager;
+
+  private transient long currentWindowId;
+
+  private transient boolean isInSafeMode;
+
+  private transient boolean isInReconcilingMode;
+
+  private transient long reconcilingWindowId;
+
+  @AutoMetric
+  transient long numInsertsInThisWindow;
+
+  @AutoMetric
+  transient long numUpsertsInThisWindow;
+
+  @AutoMetric
+  transient long numDeletedInThisWindow;
+
+  @AutoMetric
+  transient long numUpdatesInThisWindow;
+
+  @AutoMetric
+  transient long opsErrorsInThisWindow;
+
+  @AutoMetric
+  transient long bytesWrittenInThisWindow;
+
+  @AutoMetric
+  transient long rpcErrorsInThisWindow;
+
+  @AutoMetric
+  transient long writeOpsInThisWindow;
+
+  @AutoMetric
+  transient long writeRPCsInThisWindow;
+
+  @AutoMetric
+  long totalOpsErrors = 0;
+
+  @AutoMetric
+  long totalBytesWritten = 0;
+
+  @AutoMetric
+  long totalRpcErrors = 0;
+
+  @AutoMetric
+  long totalWriteOps = 0;
+
+  @AutoMetric
+  long totalWriteRPCs = 0;
+
+  @AutoMetric
+  long totalInsertsSinceStart;
+
+  @AutoMetric
+  long totalUpsertsSinceStart;
+
+  @AutoMetric
+  long totalDeletesSinceStart;
+
+  @AutoMetric
+  long totalUpdatesSinceStart;
+
+  @InputPortFieldAnnotation
+  public final transient DefaultInputPort<KuduExecutionContext> input = new DefaultInputPort<KuduExecutionContext>()
+  {
+    @Override
+    public void process(KuduExecutionContext kuduExecutionContext)
+    {
+      if (kuduExecutionContext == null) {
+        return;
+      }
+      if ( isInSafeMode) {
+        return;
+      }
+      if ( isInReconcilingMode) {
+        if ( !isEligibleForPassivationInReconcilingWindow(kuduExecutionContext, currentWindowId)) {
+          return;
```

Comment:
The current window is a new window - one that has not been processed yet. Do we want to decide whether or not to process this record depending just on `isEligibleForPassivationInReconcilingWindow()`?